src/getPublicKeys.c

```diff
@@ -21,6 +21,55 @@ static void parsePath(read_view_t* view)
 	BIP44_PRINTF(&ctx->pathSpec);
 }
 
+static void advanceStage()
+{
+	TRACE("Advancing from stage: %d", ctx->stage);
+
+	switch (ctx->stage) {
+
+	case GET_KEYS_STAGE_INIT:
+		ctx->stage = GET_KEYS_STAGE_GET_KEYS;
+
+		if (ctx->currentPath < ctx->numPaths) {
+			// there are more paths to be received
+			// so we don't want to advance beyond GET_KEYS_STAGE_GET_KEYS
+			break;
+		}
+
+	// intentional fallthrough
+
+	case GET_KEYS_STAGE_GET_KEYS:
+		ASSERT(ctx->currentPath == ctx->numPaths);
+		ctx->stage = GET_KEYS_STAGE_NONE;
+		ui_idle(); // we are done with this key export
+		break;
+
+	case SIGN_STAGE_NONE:
+	default:
+		ASSERT(false);
+	}
+}
+
+static void advanceStageIfAppropriate()
```

Comment:
I would prefer to put the condition `if (ctx->currentPath == ctx->numPaths) ` right in the part of code where this function is invoked. I see similar conditional calls of advanceStage in the signTx.c file so it should be consistent with the existing codebase as well and IMHO more readable unless I'm missing something

Reply:
The difference (vs. signTx) is that one has to distinguish if we are in _INIT or _GET_KEY state. Essentially, the body of this function can be copied to the place where it is called (there is just one), but I like to keep the code inside UI state machines as short as possible (like, a single screen), otherwise it becomes hard to read thanks to the unbalanced macros.

Reply:
why do we actually have two functions? advanceStage seems a bit redundant to me :thinking:

Reply:
This was already refactored, there is just one function (advanceStage) now.